### PR TITLE
Added Shopping Cart to online book project.

### PR DESCRIPTION
**Pull request by ponomvrenko**
It enables to add a specific book to shopping cart.

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+                            ShoppingCart shoppingCart = new ShoppingCart();
+                            shoppingCart.setUser(user);
+                            shoppingCartRepository.save(shoppingCart);
+                            return shoppingCart;
+                        });
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDB.getCartItems().add(cartItem);
+
+        return shoppingCartMapper.toDto(shoppingCartFromDB);
```
Comment (OlexVoit) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: very long. you have to split the method

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        ShoppingCart shoppingCartFromDB =
+                shoppingCartRepository.findByUserId(userId)
+                        .orElseGet(() -> {
+                            ShoppingCart shoppingCart = new ShoppingCart();
+                            shoppingCart.setUser(user);
+                            shoppingCartRepository.save(shoppingCart);
+                            return shoppingCart;
+                        });
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
```
Comment (OlexVoit) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:49]: we cannot address the cartItemRepository from the shoppingCartService

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        return shoppingCartMapper.toDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItemById(Long userId, Long cartItemId) {
+        ShoppingCart shoppingCart = findShoppingCartByUserId(userId);
+        CartItem cartItem =
+                findCartItemByShoppingCartIdAndCartItemId(cartItemId, shoppingCart);
+        cartItemRepository.delete(cartItem);
```
Comment (OlexVoit) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:76]: i assume that you need to create a CartItemService and implement the logic associated with the CartItem

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    private CartItem findCartItemByShoppingCartIdAndCartItemId(
```
Comment (OlexVoit) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]: the same

**Review by OlexVoit — CHANGES_REQUESTED** [review]
GJ)
You need to create CartItemService and implement logic

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 2 above)
Comment (ponomvrenko) [reply src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:49]: Hi, but at the lecture we talked about these methods and as I understand it's ok to use cartItem repository inside shoppingCartService.

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java (hunk 3 above)
Comment (ponomvrenko) [reply src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:76]: Including all the shopping cart logic in one place was my priority, so I did not want to create a separate service for cart items.

Comment (ponomvrenko) [reply src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:76]: My beliefs are based on the task's conditions and Sofia's lectures.

File: src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java:1]: remove unnecessary prefix 'ROLE_' from hasRole

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/onlinebookstore/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/onlinebookstore/repository/CartItemRepository.java [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package com.example.onlinebookstore.repository;
+
+import com.example.onlinebookstore.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/repository/CartItemRepository.java:8]:
```suggestion
@Repository
public interface CartItemRepository extends JpaRepository<CartItem, Long> {
```

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    public ShoppingCartResponseDto addToShoppingCart(
+            CartItemRequestDto requestDto,
+            Long userId
+    ) {
+        CartItem cartItem = cartItemFormationWithoutShoppingCart(requestDto);
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find user by this ID: " + userId
+                ));
+        ShoppingCart shoppingCartFromDB = shoppingCartRepository.findByUserIdWithCartItems(userId)
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserIdWithCartItems(userId)
```

File: src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+    private ShoppingCart findShoppingCartByUserId(Long userId) {
+        return shoppingCartRepository.findByUserIdWithCartItems(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find shopping cart by this "
+                                + "user ID: " + userId
+                ));
+    }
+
+    private CartItem cartItemFormationWithoutShoppingCart(CartItemRequestDto requestDto) {
```
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
    private CartItem createCartItemWithoutShoppingCart(CartItemRequestDto requestDto) {
```

**Review by d1sam — CHANGES_REQUESTED** [review]
good job, but there are some change should be done!

File: src/main/java/com/example/onlinebookstore/controller/BookController.java
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:1]: it's not related to this hw.

File: src/main/java/com/example/onlinebookstore/controller/CategoryController.java
Comment (d1sam) [inline src/main/java/com/example/onlinebookstore/controller/CategoryController.java:1]: it's not related to this hw.

**Review by d1sam — CHANGES_REQUESTED** [review]
please remove changes not related to this hw.

File: src/main/java/com/example/onlinebookstore/model/CartItem.java (hunk 6 above)
Comment (vovasalyha) [reply src/main/java/com/example/onlinebookstore/model/CartItem.java]: You marked this as resolved but didn't fix it. Why?

**Review by vovasalyha — APPROVED** [review]
Approving, but if you are marking some comment as resolved without making changes please leave an explanation why you don't want to introduce such change.

File: src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java
Comment (ponomvrenko) [reply src/main/java/com/example/onlinebookstore/controller/ShoppingCartController.java:1]: Done!

File: src/main/java/com/example/onlinebookstore/model/CartItem.java (hunk 6 above)
Comment (ponomvrenko) [reply src/main/java/com/example/onlinebookstore/model/CartItem.java]: @vovasalyha my fault, I just saw that Dima gave me two identical comments, before that I thought he was only one

**Pull request merged**